File: src/main/java/de/whitefrog/froggy/repository/GraphRepository.java
package de.whitefrog.froggy.repository;

import de.whitefrog.froggy.Service;
import de.whitefrog.froggy.model.Graph;
import org.neo4j.graphdb.Label;
import org.neo4j.graphdb.Node;
import org.neo4j.graphdb.Result;

public class GraphRepository {
  private Service service;
  private Graph graph = null;

  public GraphRepository(Service service) {
    this.service = service;
  }

  public Graph create() {
    if(getGraph() != null) return graph;
    return new Graph();
  }

  public Graph getGraph() {
    if(graph != null) return graph;
    Result results = service.graph().execute("match (n:Graph) return n");
    if(results.hasNext()) {
      Node node = (Node) results.next().get("n");
      graph = new Graph();
      if(node.hasProperty(Graph.Companion.getVersion())) {
        graph.setVersion((String) node.getProperty(Graph.Companion.getVersion()));
      }
    }
    return graph;
  }
  
  public void save(Graph graph) {
    if(this.graph != null) {
      Result results = service.graph().execute("match (n:Graph) return n");
      Node node = (Node) results.next().get("n");
      node.setProperty(Graph.Companion.getVersion(), graph.getVersion());
    } else {
      Node node = service.graph().createNode(Label.label("Graph"));
      node.setProperty(Graph.Companion.getVersion(), graph.getVersion());
      this.graph = graph;
    }
  }
}

File: base/src/main/java/de/whitefrog/frogr/repository/GraphRepository.kt
package de.whitefrog.frogr.repository

import de.whitefrog.frogr.Service
import de.whitefrog.frogr.model.Graph
import org.neo4j.graphdb.Label
import org.neo4j.graphdb.Node

class GraphRepository(private val service: Service) {
  private var graph: Graph? = null

  fun create(): Graph? {
    return if (getGraph() != null) graph else Graph()
  }

  fun getGraph(): Graph? {
    if (graph != null) return graph
    val results = service.graph().execute("match (n:Graph) return n")
    if (results.hasNext()) {
      val node = results.next()["n"] as Node
      graph = Graph()
      if (node.hasProperty(Graph.Version)) {
        graph!!.version = node.getProperty(Graph.Version) as String
      }
    }
    return graph
  }

  fun save(graph: Graph) {
    if (this.graph != null) {
      val results = service.graph().execute("match (n:Graph) return n")
      val node = results.next()["n"] as Node
      if (graph.version != null) node.setProperty(Graph.Version, graph.version)
    } else {
      val node = service.graph().createNode(Label.label("Graph"))
      if (graph.version != null) node.setProperty(Graph.Version, graph.version)
      this.graph = graph
    }
  }
}
